types/src/params.rs

```diff
@@ -408,6 +416,248 @@ impl<'a> Id<'a> {
+///
+/// fn main() {
+///     let mut builder = UnnamedParamsBuilder::new();
+///     builder.insert("param1");
+///     builder.insert(1);
+///     let params = builder.build();
+///
+///     // Use RPC parameters...
+/// }
+/// ```
+#[derive(Debug)]
+pub struct UnnamedParamsBuilder(ParamsBuilder);
+
+impl UnnamedParamsBuilder {
+	/// Construct a new [`UnnamedParamsBuilder`].
+	pub fn new() -> Self {
+		Self(ParamsBuilder::new('[', ']'))
+	}
+
+	/// Insert a plain value into the builder.
+	pub fn insert<P: Serialize>(&mut self, value: P) -> Result<(), serde_json::Error> {
+		self.0.insert(value)
+	}
+
+	/// Finish the building process and return a JSON compatible string.
+	pub fn build(self) -> UnnamedParams {
+		UnnamedParams(self.0.build())
+	}
+}
+
+/// Unnamed RPC parameters stored as a JSON Array object `[ value0, value1, .., valueN ]`.
+#[derive(Clone, Debug)]
+pub struct UnnamedParams(String);
+
+impl ToRpcParams for UnnamedParams {
+	fn to_rpc_params(self) -> Result<Option<Box<RawValue>>, serde_json::Error> {
+		RawValue::from_string(self.0).map(|res| Some(res))
+	}
+}
+
+/// Marker trait for types that can be serialized as JSON compatible strings.
+///
+/// This trait ensures the correctness of the RPC parameters.
+///
+/// # Note
+///
+/// Please consider using the [`UnnamedParamsBuilder`] and [`NamedParamsBuilder`] than
+/// implementing this trait.
+///
+/// # Examples
+///
+/// - Implementation for hard-coded strings
+///
+/// ```rust
+///
+/// use jsonrpsee_types::ToRpcParams;
+/// use serde_json::value::RawValue;
+///
+/// struct ManualParam;
+///
+/// impl ToRpcParams for ManualParam {
+///     fn to_rpc_params(self) -> Result<Option<Box<RawValue>>, serde_json::Error> {
+///         // Manually define a valid JSONRPC parameter.
+///         RawValue::from_string("[1, \"2\", 3]".to_string()).map(|res| Some(res))
+///     }
+/// }
+/// ```
+///
+/// - Implementation for JSON serializable structures
+///
+/// ```rust
+/// use jsonrpsee_types::ToRpcParams;
+/// use serde_json::value::RawValue;
+/// use serde::Serialize;
+///
+/// #[derive(Serialize)]
+/// struct SerParam {
+///     param_1: u8,
+///     param_2: String,
+/// };
+///
+/// impl ToRpcParams for SerParam {
+///     fn to_rpc_params(self) -> Result<Option<Box<RawValue>>, serde_json::Error> {
+/// 		let s = String::from_utf8(serde_json::to_vec(&self)?).expect("Valid UTF8 format");
+///         RawValue::from_string(s).map(|res| Some(res))
+///     }
+/// }
+/// ```
+pub trait ToRpcParams {
+	/// Consume and serialize the type as a JSON raw value.
+	fn to_rpc_params(self) -> Result<Option<Box<RawValue>>, serde_json::Error>;
+}
+
+/// Initial number of parameters in a batch request.
+const BATCH_PARAMS_NUM_CAPACITY: usize = 4;
+
+/// Parameter builder that serializes RPC parameters to construct a valid batch parameter.
+/// This is the equivalent of chaining multiple RPC requests.
+#[derive(Debug)]
+pub struct BatchParamsBuilder<'a>(Vec<(&'a str, Option<Box<RawValue>>)>);
```

Comment:
hmm, can we call this `BatchRequestBuilder` instead as it also contains the `method name` which isn't a `param` according to JSON-RPC spec :P

sorry, I'm a bit damaged after reading the JSON-RPC spec a few times :)

Reply:
`BatchRequestBuilder` sounds good to me, have updated it!